File: 0114 Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        Stack<TreeNode> st = new Stack<>();
        pushStack(st, root);
        st.pop();
        while(!st.isEmpty()) {
            root.left = null;
            root.right = st.pop();
            root = root.right;
        }
    }
    public void pushStack(Stack<TreeNode> st, TreeNode curr) {
        if (curr == null) return;
        pushStack(st, curr.right);
        pushStack(st, curr.left);
        st.push(curr);
    }
}
/* faster solution
class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        TreeNode l = root.left;
        TreeNode r = root.right;
        flatten(root.left);
        flatten(root.right);
        
        root.left = null;
        root.right = l;
        
        while (root.right != null) 
            root = root.right;  
        
        root.right = r;
        
    }
}
*/
/*
class Solution {
    public void flatten(TreeNode root) {
        if (root != null) {
            helper(root);
        }
    }
    public TreeNode helper(TreeNode curr) {
        if (curr.left == null && curr.right == null) return curr;
        TreeNode l = curr.left;
        TreeNode r = curr.right;
        
        TreeNode next = (l == null) ? r : l;
        TreeNode lEnd = (l == null) ? curr : helper(l);
        TreeNode rEnd = (r == null) ? lEnd : helper(r);
        lEnd.right = r;
        curr.left = null;
        curr.right = next;
        return rEnd;
    }
}
*/

File: 0114 Flatten Binary Tree to Linked List.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def flatten(self, root: TreeNode) -> None:
        if not root: return
        
        l, r = root.left, root.right
        self.flatten(root.left)
        self.flatten(root.right)
        root.left, root.right = None, l
        
        while root.right:
            root = root.right
        root.right = r
        
''' Morris Traversal
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def flatten(self, root: TreeNode) -> None:
        while root:
            if not root.left:
                root = root.right
            else:
                tmp = root.right
                root.right = root.left
                root.left = None
                root = root.right
                r = root
                
                while r.right:
                    r = r.right
                    
                r.right = tmp
'''